src/libsyntax/parse/parser.rs

```diff
@@ -2700,6 +2700,17 @@ impl<'a> Parser<'a> {
                 let (span, e) = self.interpolated_or_expr_span(e)?;
                 (span, self.mk_unary(UnOp::Not, e))
             }
+            // fix issue #41679: Suggest `!` for bitwise negation when encountering a `~`
+            token::Tilde => {
+                self.bump();
+                let e = self.parse_prefix_expr(None);
+                let (span, e) = self.interpolated_or_expr_span(e)?;
+                let mut err = self.diagnostic().struct_span_err(lo.to(span),
+                        "`~` can not be used as an unary operator");
+                err.help("use `!` instead of `~` if you meant to bitwise negation");
```

Comment:
Could you also add `err.span_label(span_of_tilda, "did you mean `!`?")` and use `span_of_tilda` for the primary error as well?